book/src/exec.md

```diff
@@ -0,0 +1,55 @@
+
+#### Intermediate notes
+An intermediate note is a note with a simplified sending VP. This VP is specific to the intent and contains only the rules that are necessary to satisfy for the transaction to happen. All of the note fields stay the same except:
+- sending VP -> intept-specific VP
+- owner address -> intent-specific owner address (sending VP is used to calculate the owner address, so changing the VP implies the change of the owner address)
+
+The creation of the intermediate note can be seen as sending a note to a new address with intent-specific VP.
+
+Note: as the *nullifier key* **nk** is also used to calculate the address, so the VP doesn't solely define the address
+
+Note: The content of intent-specific VPs in known to the solver because the solver will create the final transaction and some of the proofs
+
+Note: intermediate notes will not be published
+
+### Solve
+We are considering the model when a solver makes one step at a time and sends the result to the next solver. In practice, the solver can send the result to themselves and continue solving if they have the intent to make the next step. It would be nice to merge the steps into one when possible, but for simplicity we ignore this detail here
+
+Note: each solver must check the proofs they receive from other nodes
+
+When a solver has two intents that can be matched together, they match the intents by spending the old notes and creating new notes. Solvers have the authority to spend and create the intermediate notes they receive and produce the proofs required to perform the action. Solvers also know the content of intent-specific VPs and receving VPs of the users (necessary to be able to satisfy them). Solvers don't know the identities of the users. Created notes are added in a local MT tree that will be published later along with the settled transaction.
+
+If a solver has two intents that can be matched together, two cases are possible:
+1. The match is partial, partial transaction and a new intent are created. In this case the solver appends the data they produced to the new intent and gossips the new intent to the next node
+2. The match is final, the solver publishes the final transaction as described below
+
+Currently we assume a simpler model where only one solver can match n-party bartering intents (no partial solving) and an intermediate tx cannot be published on the blockchain as VPs will not allow that (balance issue). In the partial solving case partial transactions cannot be published on the blockchain as there will be a VP/other restriction mechanism blocking that.
```

Comment:
currently, as in, in the current implementation?

Reply:
Yes

Reply:
@joebebel how difficult would this be to change?

Reply:
I don't think it'll be very hard, but it's not so clear to me what the best way to do it is. One simple solution is to bring back homomorphic balance commitments and allow token VPs to "opt in". Then for any token that only balances in the usual way, there's no issues with partial solving or private solving, and we can discuss further how to allow it for other tokens.